src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+
+    const size_t maxNumberOfMazeSteps = 2 * std::sqrt(rows * columns);
+
+    std::stack<arma::ivec> cellsInPath;
+
+    arma::ivec startingPoint = {arma::randi(arma::distr_param(0, rows-1)), arma::randi(arma::distr_param(0, columns-1))};
+
+    cellsInPath.push(startingPoint);
+
+    size_t numberOfMazeSteps = 0;
+ 
+    // loop till stack is empty
+    while (!cellsInPath.empty())
+    {
+        // Pop a vertex from the stack
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+
+        ++numberOfMazeSteps;
+
+        // Break for Max number of maze steps.
+        if(numberOfMazeSteps > maxNumberOfMazeSteps)
+          break;
+ 
+        // Mark the particular row and column as visited.
+        visited(lastCell(0),lastCell(1)) = 1;
+
+        // If visitable nodes are present from that node.
+        bool visitableNodes = false;
+
+        // Randomly choose a direction and check if the adjacent node in 
+        // that direction is visited or not.
+        for (size_t counter = 0; counter < 10 ; ++counter)
+        {    
+            size_t index = arma::randi(arma::distr_param(0, 3));
+            arma::ivec nextCell = arma::ivec({lastCell(0) + moves[index](0), {lastCell(1) + moves[index](1)}});
+            
+            // Check if particular row and column index are out of bounds of maze.
+            bool outOfBounds = nextCell(0)>= maze.n_rows || nextCell(1)>= maze.n_cols
+        || nextCell(0) < 0 || nextCell(1) < 0 ;
+            if (!outOfBounds && !visited(nextCell(0),nextCell(1))) 
+            {
+                cellsInPath.push(lastCell);
+                cellsInPath.push(nextCell);
+                break;
+            }
+        }
+    }
+
+    // Pop last ({row,column}) from the stack
+      arma::ivec goalCell = cellsInPath.top();
+      cellsInPath.pop();
+
+    // Set the ({row,column}) asa goal cell
+      maze(goalCell(0),goalCell(1)) = 1;
+
+    // Set all other cells in stack to 0 
+    while (!cellsInPath.empty())
+    {
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+      
+        maze(lastCell(0),lastCell(1)) = 0;
+    }
+
+    // Randomly set other cells to -1(wall) or 0(path)
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) !=0 && maze(row,col) !=1 )
+          {
+            maze(row,col) = arma::randi(arma::distr_param(-1, 0));
+          }
+      }
+    }
+
+    // Store goal cell
+    goal = arma::vec({double(goalCell(0)),double(goalCell(1))});
+
+    // Store all the path cells as potential starting points
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) == 0)
+          {
+            startingPoints.push_back(arma::vec({double(row),double(col)}));
+          }
+      }
+    }
```

Comment:
See comment regarding STL and arma functions

Reply:
Also the spacing is off

Reply:
Another question. What's stopping us from using `arma::mat` as `startingPoints`. That way each `arma::vec` will be stored in column based index.

Reply:
Can you please explain this ?

Reply:
See previous comment reg STL, you can apply the same here.

Reply:
I was getting some error on this ! let me check that out and revert back to you.

Reply:
> Another question. What's stopping us from using `arma::mat` as `startingPoints`. That way each `arma::vec` will be stored in column based index.

I am not sure if i can understand what you are suggesting. 
I have used startingpoints as vector because we dont know the count of number of zeroes in whole matrix before hand. Therefor i loop through matrix and push_back all the valid starting points. Moreover, it helps me in selecting a random starting point. Which i do using `randi()` when i select the initial state in `InitialSample()`.

Reply:
Nvm, keep the original format. Just use STL instead of blind loops.

Reply:
@jonpsy  using std::generate or arma::for_each or arma::transform i am not able to access the row and column of the 0th cell. I think for this for loop seems ok.